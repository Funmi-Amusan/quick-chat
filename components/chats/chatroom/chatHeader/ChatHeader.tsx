import { FontAwesome } from '@expo/vector-icons';
import Ionicons from '@expo/vector-icons/Ionicons';
import { useTheme } from '@react-navigation/native';
import { router } from 'expo-router';
import { useState } from 'react';
import { View, Text, Image, TouchableOpacity } from 'react-native';

import ChatHeaderSkeleton from './ChatHeaderSkeleton';
import SearchInput from '../../shared/SearchInput';

import { ImageAssets } from '~/assets';
import { formatMomentAgo } from '~/lib/helpers';
import { ChatPartner } from '~/lib/types';

const ChatHeader = ({
  chatPartner,
  setSearchString,
  searchString,
  isLoading,
}: {
  chatPartner: ChatPartner | undefined;
  setSearchString: (arg: string) => void;
  searchString: string;
  isLoading: boolean;
}) => {
  const { dark } = useTheme();

  const [isSearchActive, setIsSearchActive] = useState(false);

  if (isLoading) {
    return <ChatHeaderSkeleton />;
  }

  return (
    <>
      <View className="border-b border-white/30 bg-body-light/20 px-4 py-2 dark:bg-greyBg-dark/20">
        {isSearchActive ? (
          <SearchInput
            setSearchString={setSearchString}
            searchString={searchString}
            closeSearchModal={() => setIsSearchActive(false)}
          />
        ) : (
          <View className="flex-row items-center gap-2">
            <TouchableOpacity testID='back-icon' className="px-2" onPress={() => router.back()}>
              <FontAwesome name="chevron-left" size={14} color={dark ? '#ffffff' : '#000000'} />
            </TouchableOpacity>
            <View className=" flex-1 flex-row items-center justify-between ">
              <View>
                <Image source={ImageAssets.avatar} className="mr-3 h-10 w-10 rounded-full" />
                <View>
                  <Text className="text-lg font-semibold text-title-light dark:text-title-dark">
                    {chatPartner?.username}
                  </Text>
                  <View className="mt-0.5 flex-row items-center text-greyText-light dark:text-greyText-dark">
                    {chatPartner?.isActive ? (
                      <>
                        <View className="mr-1 h-2 w-2 rounded-full bg-green-500" />
                        <Text className="text-xs text-gray-600">Active now</Text>
                      </>
                    ) : (
                      <Text className="text-xs text-greyText-light dark:text-greyText-dark">
                        {chatPartner?.lastActive
                          ? `Left Chatroom ${formatMomentAgo(chatPartner.lastActive)}`
                          : 'Offline'}
                      </Text>
                    )}
                  </View>
                </View>
              </View>
              <TouchableOpacity testID='search-icon' onPress={() => setIsSearchActive(true)} className="">
                <Ionicons name="search-outline" size={24} color="white" />
              </TouchableOpacity>
            </View>
          </View>
        )}
      </View>
    </>
  );
};

export default ChatHeader;
