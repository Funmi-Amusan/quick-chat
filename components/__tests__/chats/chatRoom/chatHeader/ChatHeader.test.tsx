import { fireEvent, render } from '@testing-library/react-native';
import { router } from 'expo-router';
import React from 'react';
import { TouchableOpacity } from 'react-native';

import ChatHeader from '~/components/chats/chatroom/chatHeader/ChatHeader';

jest.mock('expo-router', () => {
  // --- Re-create the stable mock functions for router methods for this file ---
  // We define them again here so they are scoped to this mock factory
  const mockBack = jest.fn();
  const mockPush = jest.fn();
  const mockReplace = jest.fn();
  const mockSetParams = jest.fn();
  // Add other router methods you need to mock in this test file

  // --- Mock the useRouter hook for this test file ---
  // It returns the object with stable mock function instances defined above
  const useRouter = () => ({
    back: mockBack,
    push: mockPush,
    replace: mockReplace,
    setParams: mockSetParams,
    // ... return other mock methods
  });

  // *** Attach the mock function instances to the mock useRouter function itself for easy access ***
  useRouter.mockBack = mockBack;
  useRouter.mockPush = mockPush;
  useRouter.mockReplace = mockReplace;
  useRouter.mockSetParams = mockSetParams;
  // Attach other mock functions here

  // --- Provide a mock implementation for the <Link> component for this test file ---
  const Link = (props: any) => {
    const { children, href, asChild, ...restProps } = props; // Include asChild if you use it
    // Render a simple TouchableOpacity or View that passes through props and href
    // Use a testID to find it in tests and data-href for debugging
    return React.createElement(
      TouchableOpacity, // Using TouchableOpacity might be closer to Link's behavior
      {
        ...restProps, // Spread other props like style, accessibilityLabel, onPress etc.
        'data-href': href, // Useful for debugging
        testID: props.testID || `LinkMock-${href}`, // Use href in the testID for uniqueness if testID is not provided
        // Note: If your component attaches an onPress to Link and you want to test that,
        // the fireEvent.press on this TouchableOpacity mock will trigger it.
      },
      children // Render the children inside the mock Link
    );
  };

  // Mock other expo-router exports if needed in this specific test file
  const useSegments = () => [];
  const useLocalSearchParams = () => ({});

  // --- Return all the mocked exports needed for this test file ---
  return {
    // Return the mock useRouter function (with attached mocks)
    useRouter,

    // Return the mock Link component
    Link,

    // Include other mocks needed in this specific test file
    useSegments,
    useLocalSearchParams,
    // ... other mocked exports
  };
});

test('displays chatList item correctly', async () => {
  const replyPreviewComponent = render(
    <ChatHeader
      chatPartner={undefined}
      setSearchString={(e) => {}}
      searchString=""
      isLoading={false}
    />
  );

  expect(replyPreviewComponent.toJSON()).toMatchSnapshot();
});
test('renders search input when search is active', () => {
  const { getByTestId, getByPlaceholderText } = render(
    <ChatHeader
      chatPartner={{
        username: 'TestUser',
        isActive: true,
        isLoggedIn: true,
        id: '123',
        isTyping: { isTyping: false },
        lastActive: new Date().getTime(),
      }}
      setSearchString={() => {}}
      searchString=""
      isLoading={false}
    />
  );

  const searchButton = getByTestId('search-icon');
  fireEvent.press(searchButton);

  expect(getByPlaceholderText('Search Chats')).toBeTruthy();
});

test('renders loading skeleton when isLoading is true', () => {
  const { getByTestId } = render(
    <ChatHeader chatPartner={undefined} setSearchString={() => {}} searchString="" isLoading />
  );

  expect(getByTestId('chat-header-skeleton')).toBeTruthy();
});

test('displays active status correctly when user is online', () => {
  const { getByText } = render(
    <ChatHeader
      chatPartner={{
        username: 'TestUser',
        isActive: true,
        isLoggedIn: true,
        id: '123',
        isTyping: { isTyping: false },
        lastActive: new Date().getTime(),
      }}
      setSearchString={() => {}}
      searchString=""
      isLoading={false}
    />
  );

  expect(getByText('Active now')).toBeTruthy();
});

test('displays offline status with last active time', () => {
  const lastActive = new Date(Date.now() - 3600000).toISOString(); // 1 hour ago
  const { getByText } = render(
    <ChatHeader
      chatPartner={{
        username: 'TestUser',
        isActive: false,
        isLoggedIn: true,
        id: '123',
        isTyping: { isTyping: false },
        lastActive: new Date().getTime(),
      }}
      setSearchString={() => {}}
      searchString=""
      isLoading={false}
    />
  );

  expect(getByText(/Left Chatroom/)).toBeTruthy();
});

test('displays offline status when lastActive is not available', () => {
  const { getByText } = render(
    <ChatHeader
      chatPartner={{
        username: 'TestUser',
        isActive: false,
        isLoggedIn: true,
        id: '123',
        isTyping: { isTyping: false },
        lastActive: null,
      }}
      setSearchString={() => {}}
      searchString=""
      isLoading={false}
    />
  );

  expect(getByText('Offline')).toBeTruthy();
});

// test('calls router.back when back button is pressed', () => {
//   // const mockRouter = jest.spyOn(router, 'back');
//   const { getByTestId} = render(
//     <ChatHeader
//       chatPartner={{
//         username: 'TestUser',
//         isActive: true,
//         isLoggedIn: true,
//         id: '123',
//         isTyping: { isTyping: false },
//         lastActive: new Date().getTime(),
//       }}
//       setSearchString={() => {}}
//       searchString=""
//       isLoading={false}
//     />
//   );

//   const backButton = getByTestId('back-icon');
//   fireEvent.press(backButton);
// debug()
//   // expect(mockRouter).toHaveBeenCalled();
// });
