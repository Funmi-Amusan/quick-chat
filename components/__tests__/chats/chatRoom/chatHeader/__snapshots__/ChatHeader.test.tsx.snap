// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays chatList item correctly 1`] = `
<View
  className="border-b border-white/30 bg-body-light/20 px-4 py-2 dark:bg-greyBg-dark/20"
>
  <View
    className="flex-row items-center gap-2"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="back-icon"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#000000",
              "fontSize": 14,
            },
            undefined,
            {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
    <View
      className=" flex-1 flex-row items-center justify-between "
    >
      <View>
        <Image
          className="mr-3 h-10 w-10 rounded-full"
          source={
            {
              "testUri": "../../../assets/images/Girl.png",
            }
          }
        />
        <View>
          <Text
            className="text-lg font-semibold text-title-light dark:text-title-dark"
          />
          <View
            className="mt-0.5 flex-row items-center text-greyText-light dark:text-greyText-dark"
          >
            <Text
              className="text-xs text-greyText-light dark:text-greyText-dark"
            >
              Offline
            </Text>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="search-icon"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "white",
                "fontSize": 24,
              },
              undefined,
              {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
